docs/multi_credential_assignment.md

```diff
@@ -0,0 +1,163 @@
+
+`GET /api/v2/job_templates/N/credentials/`
+
+Users can associate and disassociate credentials using `POST` requests to this
+interface, similar to the behavior in the now-deprecated `extra_credentials`
+endpoint:
+
+```
+POST /api/v2/job_templates/N/credentials/ {'associate': true, 'id': X'}
+POST /api/v2/job_templates/N/credentials/ {'disassociate': true, 'id': Y'}
+```
+
+Under this model, a JobTemplate is considered valid even when it has _zero_
+Credentials assigned to it.
+
+Launch Time Considerations
+--------------------------
+
+Prior to this change, JobTemplates had a configurable attribute,
+`ask_credential_on_launch`.  This value was used at launch time to determine
+which missing credential values were necessary for launch - this was primarily
+used as a mechanism for users to specify a Machine/SSH credential to satisfy 
+the minimum Credential requirement.
+
+Under the new unified Credential list model, this attribute still exists, but it
+is no longer bound to a notion of "requiring" a Credential.  Now when
+`ask_credential_on_launch` is `True`, it signifies that users may (if they
+wish) specify a list of credentials at launch time to override those defined on
+the JobTemplate:
+
+`POST /api/v2/job_templates/N/launch/ {'credentials': [A, B, C]}`
+
+If `ask_credential_on_launch` is `False`, it signifies that custom `credentials`
+provided in the payload to `POST /api/v2/job_templates/N/launch/` will be
+ignored.
+
+Under this model, the only purpose for `ask_credential_on_launch` is to signal
+that API clients should prompt the user for (optional) changes at launch time.
+
+Backwards Compatability Concerns
+--------------------------------
+A variety of API clients rely on now-deprecated mechanisms for Credential
+retrieval and assignment, and those are still supported in a backwards
+compatible way under this new API change.  Requests to update
+`JobTemplate.credential` and `JobTemplate.vault_credential` will still behave
+as they did before:
+
+`PATCH /api/v2/job_templates/N/ {'credential': X, 'vault_credential': Y}`
+
+Under this model, when a JobTemplate with multiple vault Credentials is updated
+in this way, the new underlying list will _only_ contain the single Vault
+Credential specified in the deprecated request.
+
+`GET` requests to `/api/v2/job_templates/N/` and `/api/v2/jobs/N/`
+have traditionally included a variety of metadata in the response via
+`related_fields`:
+
+```
+{
+    "related": {
+        ...
+        "credential": "/api/v2/credentials/1/",
+        "vault_credential": "/api/v2/credentials/3/",
+        "extra_credentials": "/api/v2/job_templates/5/extra_credentials/",
+    }
+}
+```
+
+...and `summary_fields`:
+
+```
+{
+    "summary_fields": {
+        "credential": {
+            "description": "",
+            "credential_type_id": 1,
+            "id": 1,
+            "kind": "ssh",
+            "name": "Demo Credential"
+        },
```

Comment:
how does this still return if `JobTemplate.credential` is now an integer? I'm trying to identify what kind of special-case code you have to furnish this.

Reply:
The code that generates `summary_fields` loops over `.credentials.all()` and organizes by `CredentialType.kind`:

https://github.com/ansible/awx/pull/595/files#diff-a81324c523b41de7296fdd5ff9063d10R2560